### ISSUE TITLE
OSSL_PROVIDER_available returns True for providers which are not listed in config file.

### ISSUE THREAD

**Issue by Pankul94** [openssl/openssl#19374]
I am using Openssl-3 with FIPS provider and want to confirm that my algorithms are provided by the fips module. I have written a small C script for this:
```C
#include <stdio.h>
#include <openssl/provider.h>
#include <openssl/evp.h>
#include <string.h>

int main(int argc, char *argv[])
{
    if (argc < 3){
        printf("Usage: ./canifips <digest algorithm> <provider-name>");
        exit(1);
    }
    EVP_MD_CTX * mdctx;
    EVP_MD * algo;
    mdctx = EVP_MD_CTX_new();
    char prov[] = "provider=";
    strcat(prov,argv[2]);
    algo = EVP_MD_fetch(NULL,argv[1],prov);
    if (algo == NULL){
        printf("cannot fetch the algorithm");
        exit(1);
    }
    printf("EVP_DigestInit_ex(mdctx,algo,NULL): %d\n", EVP_DigestInit_ex(mdctx,algo,NULL));
    printf("%s OSSL_PROVIDER_available: %d\n",argv[2],OSSL_PROVIDER_available(NULL,argv[2]));
    printf(OSSL_PROVIDER_get0_name(EVP_MD_get0_provider(EVP_MD_CTX_get0_md(mdctx))));
}

```
The script takes two command line arguments, digest-algorithm and provider.

Here are the contents of my conf file:
```
config_diagnostics = 1
openssl_conf = openssl_init
.include /Openssl/build/ssl/fipsmodule.cnf
[openssl_init]
providers = provider_sect
[provider_sect]
fips = fips_sect
base = base_sect
[base_sect]
activate = 1
```

The output for `./canifips MD5 fips`:
```
cannot fetch the algorithm%
```
which makes sense as MD5 is not permitted by FIPS. 

Also running `./canifips MD5 default` results in:
```
cannot fetch the algorithm
```
which is expected, as default provider is not present in my config file.

`./canifips SHA256 fips`
```
EVP_DigestInit_ex(mdctx,algo,NULL): 1
fips OSSL_PROVIDER_available: 1
fips
```
which is again expected.

Since OSSL_PROVIDER_available returns 1 when a provider is loaded, I modified the code and moved the API call before EVP_MD_FETCH():
```C
#include <stdio.h>
#include <openssl/provider.h>
#include <openssl/evp.h>
#include <string.h>

int main(int argc, char *argv[])
{
    if (argc < 3){
        printf("Usage: ./canifips <digest algorithm> <provider-name>");
        exit(1);
    }
    printf("%s OSSL_PROVIDER_available: %d\n",argv[2],OSSL_PROVIDER_available(NULL,argv[2]));
    EVP_MD_CTX * mdctx;
    EVP_MD * algo;
    mdctx = EVP_MD_CTX_new();
    char prov[] = "provider=";
    strcat(prov,argv[2]);
    algo = EVP_MD_fetch(NULL,argv[1],prov);
    if (algo == NULL){
        printf("cannot fetch the algorithm");
        exit(1);
    }
    printf("EVP_DigestInit_ex(mdctx,algo,NULL): %d\n", EVP_DigestInit_ex(mdctx,algo,NULL));
    printf("%s OSSL_PROVIDER_available: %d\n",argv[2],OSSL_PROVIDER_available(NULL,argv[2]));
    printf(OSSL_PROVIDER_get0_name(EVP_MD_get0_provider(EVP_MD_CTX_get0_md(mdctx))));
}

```

Now when I run `./canifips MD5 default`:
```
default OSSL_PROVIDER_available: 1
EVP_DigestInit_ex(mdctx,algo,NULL): 1
default OSSL_PROVIDER_available: 1
default
```
I expected the output to be `cannot fetch the algorithm` but turns out openssl is able to load MD5 from default provider. Does the call to `OSSL_PROVIDER_available` load the library as well?

And I have not touched my config file during these changes, my config file is the same.

**Comment by paulidale** [comment]
This bit of code overwrites memory and everything after is thus suspect.
```c
    char prov[] = "provider=";
    strcat(prov,argv[2]);
```

It's not clear why you are seeing the behaviour you are.  OSSL_PROVIDER_available() can load the fallback providers (i.e. _default_) but it shouldn't with your configuration file.

**Comment by t8m** [comment]
We have some missing `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG)` call somewhere.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input

**Comment by paulidale** [comment]
I think this should be closed.   There is a blatant buffer overrun corrupting things before the supposed problem...
